### PR TITLE
mgr/cephadm: allow binding to loopback for rgw daemons

### PR DESCRIPTION
**Pull request by rkachach**
Fixes: https://tracker.ceph.com/issues/57304

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by rkachach** [comment]
jenkins retest this please

**Comment by adk3798** [comment]
jenkins test make check

File: src/pybind/mgr/cephadm/schedule.py [hunk 1]
```diff
@@ -362,6 +363,11 @@ def expand_candidates(ls: List[DaemonPlacement], num: int) -> List[DaemonPlaceme
     def find_ip_on_host(self, hostname: str, subnets: List[str]) -> Optional[str]:
         for subnet in subnets:
             ips: List[str] = []
+            # following is to allow loopback interfaces for both ipv4 and ipv6
+            if ipaddress.ip_address('127.0.0.1') in ipaddress.ip_network(subnet):
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/schedule.py:367]: interesting, I didn't know you could just do a `<ip> in <network>` check to see if the ip was in a certain subnet. We could maybe use similar to address https://tracker.ceph.com/issues/57060#note-7

**Review by adk3798 — APPROVED** [review]

File: src/pybind/mgr/cephadm/schedule.py (hunk 1 above)
Comment (rkachach) [reply src/pybind/mgr/cephadm/schedule.py:367]: > interesting, I didn't know you could just do a `<ip> in <network>` check to see if the ip was in a certain subnet. We could maybe use similar to address https://tracker.ceph.com/issues/57060#note-7

Yeah, it's very handy (and clean) to perform the check. The code behind it is the following. I think we can use it to solve the issue you referenced :+1: 

```
    def __contains__(self, other):
        # always false if one is v4 and the other is v6.
        if self._version != other._version:
            return False
        # dealing with another network.
        if isinstance(other, _BaseNetwork):
            return False
        # dealing with another address
        else:
            # address
            return other._ip & self.netmask._ip == self.network_address._ip

```

**Comment by rkachach** [comment]
jenkins test make check

**Comment by rkachach** [comment]
jenkins test make check arm64

**Comment by rkachach** [comment]
jenkins test windows

**Comment by rkachach** [comment]
jenkins test windows

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-08-30_17:06:47-orch:cephadm-wip-adk-testing-2022-08-29-1644-distro-default-smithi/
and reruns https://pulpito.ceph.com/adking-2022-09-01_12:53:01-orch:cephadm-wip-adk-testing-2022-08-29-1644-distro-default-smithi/

Failures tracked by:

- https://tracker.ceph.com/issues/57287
- https://tracker.ceph.com/issues/57374
- https://tracker.ceph.com/issues/57371
- https://tracker.ceph.com/issues/56573
- https://tracker.ceph.com/issues/55347

the only one that is new and surprising on this run were the iscsi test failures (https://tracker.ceph.com/issues/57371). As a result, I'm not going to merge any PRs in this run that seem related to iscsi. Other PRs I think will still be okay to merge though.

**Review by adk3798 — APPROVED** [review]

**Pull request merged**